ykrt/src/compile/jitc_yk/codegen/x64/rev_analyse.rs

```diff
@@ -1,142 +1,265 @@
+                {
+                    match inst {
+                        Inst::Param(pinst) => {
+                            if let VarLocation::Register(reg) = VarLocation::from_yksmp_location(
+                                self.m,
+                                iidx,
+                                self.m.param(pinst.paramidx()),
+                            ) {
+                                self.push_reg_hint_fixed(jump_op.unpack(self.m), reg);
+                            }
                         }
+                        _ => break,
                     }
-                    Inst::Store(x) => {
-                        if self.an_store(iidx, x) {
-                            continue;
-                        }
+                }
+            }
+            TraceKind::HeaderAndBody => {
+                // We don't care where the register allocator ends at the end of the header, so we
+                // don't propagate backwards from `TraceHeaderEnd`.
+            }
+            TraceKind::Sidetrace => {
+                let vlocs = self.m.root_entry_vars();
+                // Side-traces don't have a trace body since we don't apply loop peeling and thus use
+                // `trace_header_end` to store the jump variables.
+                debug_assert_eq!(vlocs.len(), self.m.trace_header_end().len());
+
+                for (vloc, jump_op) in vlocs.iter().zip(self.m.trace_header_end()) {
+                    if let VarLocation::Register(reg) = *vloc {
+                        self.push_reg_hint_fixed(jump_op.unpack(self.m), reg);
                     }
-                    Inst::TraceHeaderEnd => {
-                        self.an_header_end();
+                }
+            }
+        }
+
+        // ...and then we perform the rest of the reverse analysis.
+        let mut iter = self.m.iter_skipping_insts().rev();
+        match self.m.tracekind() {
+            TraceKind::HeaderOnly | TraceKind::Sidetrace => {
+                for (iidx, inst) in self.m.iter_skipping_insts().rev() {
+                    self.analyse(iidx, inst);
+                }
+            }
+            TraceKind::HeaderAndBody => {
+                // OPT: We could pass the index for `TraceHeaderEnd` around, perhaps in the
+                // `TraceKind` to avoid having to find it this way.
+                let mut next = iter.next();
+                while let Some((_, inst)) = next {
+                    if let Inst::TraceHeaderEnd = inst {
+                        break;
                     }
-                    Inst::TraceBodyEnd => {
-                        self.an_body_end();
+                    next = iter.next();
+                }
+
+                while let Some((iidx, inst)) = next {
+                    self.analyse(iidx, inst);
+                    next = iter.next();
+                }
+            }
+        }
+    }
+
+    /// Analyse a trace body. This must be called iff both of the following are true:
```

Comment:
would it be worth adding some debug_asserts for these preconditions then?

Reply:
I did try but the only way of doing so is via global state: I ended up removing it as it just felt like overkill (probably 10LoC, spread over the file, for something that you'd have to be fairly silly to do). I'm not sure I can think of a better way TBH...